content/posts/2023-07-24-breakdown-of-http-clients-in-elixir/index.md

```diff
@@ -0,0 +1,458 @@
+sacrifice some of the convenience provided by "higher-level" clients. It's also
+great when you know you'll have to make a lot of requests to the same host,
+since you can specify dedicated connection pools per host.
+
+### Req
+
+[Req][req] is a relatively-new kid on the block when it comes to HTTP clients in
+Elixir. It's one of my favorite Elixir libraries out there.
+
+```elixir
+Req.get!("https://api.github.com/repos/wojtekmach/req").body["description"]
+#=> "Req is a batteries-included HTTP client for Elixir."
+```
+
+It's built on top of Finch, and it takes a quite "functional-programming"
+approach to HTTP. What I mean by that is that Req revolves around a
+[`Req.Request`][req-request-docs] *data structure*, which you manipulate to add
+options, callbacks, headers, and more before making an HTTP request.
+
+```elixir
+req =
+  %Req.Request{method: :get, url: "https://github.com/..."}
+  |> Req.Request.append_request_steps(
+    put_user_agent: &Req.Steps.put_user_agent/1
+  )
+  |> Req.Request.append_response_steps(
+    decompress_body: &Req.Steps.decompress_body/1,
+    decode_body: &Req.Steps.decode_body/1
+  )
+  |> Req.Request.append_error_steps(retry: &Req.Steps.retry/1)
+
+{req, resp} = Req.Request.run_request(req)
+```
+
+Req is extensively customizable, since you can [write
+plugins][req-request-docs-plugins] for it in order to build HTTP clients that
+are tailored to your application.
+
+#### When To Use Req
+
+First and foremost, Req is *fantastic* for scripting. With the introduction of
+[`Mix.install/2`][docs-mix-install] in Elixir 1.12, using libraries in Elixir
+scripts is a breeze, and Req fits like a glove.
+
+```elixir
+Mix.install([
+  {:req, "~> 0.3.0"}
+])
+
+Req.get!("https://andrealeopardi.com").headers
+#=> [
+#=>   {"connection", "keep-alive"},
+#=>   ...
+#=> ]
+```
+
+Req is also a great fit to use in your applications. It provides a ton of
+features and plugins to use for things like encoding and decoding request
+bodies, instrumentation, authentication, and so much more. I'll take a quote
+straight from Finch's README here:
+
+> Most developers will most likely prefer to use the fabulous HTTP client
+> [Req][req] which takes advantage of Finch's pooling and provides an extremely
+> friendly and pleasant to use API.
+
+So, yeah.
+
+### httpc
+
+While Mint is the lowest-level HTTP client I know of, there's another client
+worth mentioning alongside it: [`httpc`][httpc]. `httpc` ships with the Erlang
+standard library, making it the **only** HTTP client in the ecosystem that
+*doesn't require any additional dependencies*. This is so appealing! There are
+cases where not having dependencies is a huge bonus. For example, if you're a
+library author, being able to make HTTP requests without having to bring in
+additional dependencies can be great, because those additional dependencies
+would trickle down (as transitive dependencies) to all users of your library.
```

Comment:
You can give tailwind and esbuild libraries as examples of httpc is used once or twice to download executables and then it is never used again, so no additional dependencies are really worth it.